File: android/app/src/main/kotlin/com/teamapt/customers/moniepoint/lib/LivelinessMethodHandler.kt
package com.teamapt.customers.moniepoint.lib

import android.content.Context
import android.graphics.Rect
import android.util.Size
import android.view.Surface
import androidx.camera.core.Preview
import androidx.camera.core.SurfaceRequest
import androidx.core.content.ContextCompat
import com.teamapt.livelinessdetector.LivelinessDetector
import com.teamapt.livelinessdetector.camera.CustomCamera2
import com.teamapt.livelinessdetector.events.*
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.view.TextureRegistry
import java.util.concurrent.Executor

class LivelinessMethodHandler(
        private val context: Context,
        private val textureEntry: TextureRegistry.SurfaceTextureEntry
) : MethodChannel.MethodCallHandler {
    
    private var livelinessDetector: LivelinessDetector? = null

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "create" -> initializeCamera(call, result)
            "start" -> {
                livelinessDetector?.startPreview()
                result.success("started")
            }
            "pause" -> {
                livelinessDetector?.pauseListening()
                result.success("")
            }
            "resume" -> {
                livelinessDetector?.resumeListening()
                result.success("")
            }
            "begin_capture" -> {
                livelinessDetector?.beginCapture()
                result.success("")
            }
            "end_capture" -> {
                livelinessDetector?.endCapture()
                result.success("")
            }
            "restart" -> {
                livelinessDetector?.reset()
                result.success("")
            }
            "close" -> {
                livelinessDetector?.stopCamera()
                result.success("")
            }
        }
    }

    private fun initializeCamera(methodCall: MethodCall?, reply: MethodChannel.Result) {
        val flutterSurfaceProvider = FlutterSurfaceProvider(textureEntry, ContextCompat.getMainExecutor(context))
        val customCamera = CustomCamera2(context, flutterSurfaceProvider)
        var frameSize: Rect? = null
        var previewSize: Size? = null
        val channelFrameSize = methodCall?.argument<Map<String?, Double?>>("frameSize")
        val channelPreviewSize = methodCall?.argument<Map<String?, Double?>>("previewSize")
        println("channelFrameSize $channelFrameSize")
        if(channelFrameSize != null) {
            val left = channelFrameSize["left"]
            val top = channelFrameSize["top"]
            val right = channelFrameSize["right"]
            val bottom = channelFrameSize["bottom"]

            if((left != null && left > 0)
                    && (top != null && top > 0)
                    && (right != null && right > 0)
                    && (bottom != null && bottom > 0)) {
                frameSize = Rect(left.toInt(), top.toInt(), right.toInt(), bottom.toInt())
            }
        }
        if(channelPreviewSize != null) {
            val width = channelPreviewSize["width"]
            val height = channelPreviewSize["height"]

            if(width != null && height != null) {
                previewSize = Size(width.toInt(), height.toInt())
            }
        }
        //let's check if the frameSize is supplied from the channel
        this.livelinessDetector = LivelinessDetector.Builder(customCamera)
                .setFrameSize(frameSize)
                .setCropCaptureToFrameSize(false)
                .setFaceDetectionConfidenceLevel(0.505)
                .setCompressionQuality(80)
                .setFaceDistanceToFrameThreshold(0.5)
                .setPreviewSize(previewSize ?: Size(480, 640))
                .build()
        reply.success(hashMapOf<String, Any>("cameraId" to textureEntry.id()))
    }

    /**
     *
     */
    inner class LivelinessEvenChannel: EventChannel.StreamHandler {

        private var eventSink: EventChannel.EventSink? = null

        override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
            this.eventSink = events
            livelinessDetector?.startListening {
                ContextCompat.getMainExecutor(context).execute {
                    sendEvent(it)
                }
            }
        }

        override fun onCancel(arguments: Any?) {
            eventSink = null
        }

        private fun sendEvent(cameraMotionEvent: CameraMotionEvent) {
            when(cameraMotionEvent) {
                is FirstCaptureEvent -> {
                    eventSink?.success(
                            hashMapOf<String?, Any?>(
                                    "event_type" to cameraMotionEvent.eventName,
                                    "event_data" to cameraMotionEvent.bitmap?.path
                            )
                    )
                }
                is FaceDetectedEvent -> {
                    eventSink?.success(
                            hashMapOf<String?, Any?>(
                                    "event_type" to cameraMotionEvent.eventName,
                                    "event_data" to cameraMotionEvent.bitmap?.path,
                                    "exposure" to cameraMotionEvent.exposure
                            )
                    )
                }
                is NoFaceDetectedEvent -> {
                    eventSink?.success(
                            hashMapOf<String?, Any?>(
                                    "event_type" to cameraMotionEvent.eventName,
                                    "event_data" to cameraMotionEvent.bitmap?.path
                            )
                    )
                }
                is MotionDetectedEvent -> {
                    eventSink?.success(
                            hashMapOf<String?, Any?>(
                                    "event_type" to cameraMotionEvent.eventName,
                                    "event_data" to cameraMotionEvent.bitmap?.path
                            )
                    )
                }
                is ImageOverExposed -> {
                    eventSink?.success(
                        hashMapOf<String?, Any?>(
                            "event_type" to cameraMotionEvent.eventName,
                            "event_data" to cameraMotionEvent.exposure
                        )
                    )
                }
                is ImageUnderExposed -> {
                    eventSink?.success(
                        hashMapOf<String?, Any?>(
                            "event_type" to cameraMotionEvent.eventName,
                            "event_data" to cameraMotionEvent.exposure
                        )
                    )
                }
                is DetectedFaceRectEvent -> {
                    eventSink?.success(
                        hashMapOf<String?, Any?>(
                            "event_type" to cameraMotionEvent.eventName,
                            "event_data" to """${cameraMotionEvent.faceRect.left}
                                |,${cameraMotionEvent.faceRect.top},
                                | ${cameraMotionEvent.faceRect.right},
                                |  ${cameraMotionEvent.faceRect.bottom}""".trimMargin()
                        )
                    )
                }
                is FaceNotCenteredEvent -> {
                    eventSink?.success(
                        hashMapOf<String?, Any?>(
                            "event_type" to cameraMotionEvent.eventName,
                            "event_data" to """""".trimMargin()
                        )
                    )
                }
                else -> {
                    eventSink?.success(
                            hashMapOf<String?, Any?>(
                                    "event_type" to cameraMotionEvent.eventName,
                            )
                    )
                }
            }
        }

    }

    class FlutterSurfaceProvider(
            private val surfaceTextureEntry: TextureRegistry.SurfaceTextureEntry,
            private val executor: Executor
    ) : Preview.SurfaceProvider {

        override fun onSurfaceRequested(request: SurfaceRequest) {
            val surfaceTexture = surfaceTextureEntry.surfaceTexture()
            surfaceTexture.setDefaultBufferSize(640, 480)
            request.provideSurface(Surface(surfaceTexture), executor, {})
        }
    }
}

File: ios/Runner/lib/LivelinessMethodHandler.swift
//
//  LivelinessMethodHandler.swift
//  Moniepoint
//
//  Created by Paul Okeke on 29/07/2021.
//

import Foundation
import Flutter
import UIKit



class LivelinessMethodHandler : NSObject, FlutterStreamHandler {

    private let LIVELINESS_CHANNEL: String = "moniepoint.flutter.dev/liveliness"
    private let LIVELINESS_EVENT_CHANNEL = "moniepoint.flutter.dev/liveliness/events"
    private let LIVELINESS_REGISTRAR = "moniepoint.flutter.dev/liveliness/liveliness_test"
    
    private let controller: FlutterViewController
    private var livelinessDetector: CustomCamera2?
    private var eventSink: FlutterEventSink?
    weak private var registrar: FlutterPluginRegistrar?
    
    init(controller: FlutterViewController) {
        self.controller = controller
        self.registrar = self.controller.engine?.registrar(forPlugin: LIVELINESS_REGISTRAR)
    }
    
    func registerLivelinessMethodHandler() {
        let livelinessChannel = FlutterMethodChannel(name:LIVELINESS_CHANNEL,
                                                     binaryMessenger:controller.binaryMessenger)
        
        let livelinessEventChannel = FlutterEventChannel(name: LIVELINESS_EVENT_CHANNEL,
                                                         binaryMessenger: controller.binaryMessenger)
        
        livelinessEventChannel.setStreamHandler(self)
        
        livelinessChannel.setMethodCallHandler { (call: FlutterMethodCall, reply:@escaping FlutterResult) in
            var mapResult : [String: Any?] = [:]
            print("Calling Method -> \(call.method)")
            do {
                switch call.method {
                case "create":
                    self.initializeCamera(call: call, reply: reply)
                    return
                case "close":
                    self.livelinessDetector?.stop()
                    reply([:])
                    return
                case "pause":
                    self.livelinessDetector?.pauseListening()
                    reply([:])
                    return
                case "resume":
                    self.livelinessDetector?.resumeListening()
                    reply([:])
                    return
                case "begin_capture":
                    self.livelinessDetector?.beginCapture()
                    reply([:])
                    return
                case "end_capture":
                    self.livelinessDetector?.endCapture()
                    reply([:])
                    return
                case "restart":
                    self.livelinessDetector?.reset()
                    reply([:])
                    return
                case "start":
                    reply([:])
                    return
                default:
                    return
                }
            } catch {
                print("Unexpected Error: \(error)")
            }
        }
    }
    
    func initializeCamera(call: FlutterMethodCall, reply:@escaping FlutterResult) {
        if(registrar == nil) {
            self.registrar = self.controller.engine?.registrar(forPlugin: LIVELINESS_REGISTRAR)
        }
        
        print("ARGUMENTS ==>>> \(call.arguments)")
        
        let texture: FlutterTextureRegistry = (registrar?.textures())!

        if let args = call.arguments as? Dictionary<String, Any>,
           let channelFrameSize = args["frameSize"] as? Dictionary<String?, Double?>?,
           let channelPreviewSize = args["previewSize"] as? Dictionary<String?, Double?>? {
            
            var frameSize: CGRect? = nil
            var previewSize: CGSize = CGSize(width:480, height:640)
            
            if(channelFrameSize != nil) {
                let left = channelFrameSize!["left"]
                let top = channelFrameSize!["top"]
                let right = channelFrameSize!["right"]
                let bottom = channelFrameSize!["bottom"]

                if((left != nil) && (top != nil) && (right != nil) && (bottom != nil)) {
                    frameSize = CGRect(
                        x: left!!,
                        y: (top ?? 0.0)!,
                        width: (right ?? 0.0)!,
                        height: (bottom ?? 0.0)!
                    )
                }
            }

            if(channelPreviewSize != nil) {
                let  width = channelPreviewSize!["width"]
                let height = channelPreviewSize!["height"]

                if(width != nil && height != nil) {
                    previewSize = CGSize(width: width!!, height: height!!)
                }
            }

            self.livelinessDetector = CustomCamera2(
                textureRegistry: texture, previewSize: previewSize, frameSize: frameSize)
            
            livelinessDetector?.startCamera()
            
            let cameraId = texture.register(self.livelinessDetector!)
            
            livelinessDetector?.setFrameAvailableHander {
                texture.textureFrameAvailable(cameraId)
            }
            
            print("CameraID => \(cameraId )")
            reply(["cameraId": cameraId])
        }
        else {
            
        }
    }
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        livelinessDetector?.startListening(eventHandler: { CameraMotionEvent in
            print("This is the Event \(CameraMotionEvent)")
            switch CameraMotionEvent {
            case .NoFaceDetectedEvent:
                self.eventSink?([
                    "event_type":"NoFaceDetectedEvent",
                    "event_data" :""
                ])
                return
            case .FirstCaptureEvent(path: let path):
                self.eventSink?([
                    "event_type":"FirstCaptureEvent",
                    "event_data" :path?.path
                ])
                return
            case .MotionDetectedEvent(path: let path):
                self.eventSink?([
                    "event_type":"MotionDetectedEvent",
                    "event_data" :path?.path
                ])
                return
            case .NoMotionDetectedEvent(path: let path):
                self.eventSink?([
                    "event_type": "NoMotionDetectedEvent",
                    "event_data": ""
                ])
                return
            case .FaceDetectedEvent:
                self.eventSink?([
                    "event_type":"FaceDetectedEvent",
                    "event_data" :""
                ])
                return
            case .DetectedFaceRectEvent(faceRect: let faceRect):
                self.eventSink?([
                    "event_type":"DetectedFaceRectEvent",
                    "event_data" :"\(faceRect.minX),\(faceRect.minY),\(faceRect.maxX),\(faceRect.maxY)"
                ])
                return
            case .FaceOutOfBoundsEvent:
                self.eventSink?([
                    "event_type":"FaceOutOfBoundsEvent",
                ])
                return
            }
        })
        
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        return nil
    }
    
    
}
